### ISSUE TITLE
[BUG] Have a retry on 504 gateway timeout errors when polling on a resource state

### ISSUE THREAD

**Issue by ptesny** [SAP/terraform-provider-btp#985]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### What version of the Terraform provider are you using?

1.8.0

### What version of the Terraform CLI are you using?

1.9.3

### What type of issue are you facing

bug report

### Describe the bug

Depending on the capacity pressure combined with the network distance of the btp region it may happen the BTP CLI will timeout with 504 gateway timeout error causing the abort of a terraform apply run.

The BTP resource in question (most of often seen with SWZ subscriptions and kyma runtime envs) continues to be created asynchronously as nothing happened. 

However the btp provider return an error and the whole apply run fails.

### Expected Behavior

Add more resilience  (timeout, with a number of retries, and a circuit breaker) when it comes to resources state polling scenarios

**Comment by prajin-op** [comment]
@ptesny Thank you for reporting this issue. And yes, We have observed this behaviour during the creation and deletion of Kyma runtime environment instances, which occurs due to a gateway timeout. We will discuss this internally and then proceed with the increase of timeout. We'll keep you posted.

**Comment by ptesny** [comment]
@prajin-op Thanks for the heads up. 
I think 504 errors must be mitigated by the provider itself.

However, from time to time there is this dreadful 500 error, due to probably a greater network latency (for instance running a tf script from Europe against a BTP landscape in Asia). 

Still, this has no adverse impact on the runtime creation which heads to its completion....on the BTP side.

Looks like runtime status polling with the BTP CLI APIs is not a very robust operation.

```
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [5m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [5m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m0s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m10s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m20s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m30s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [6m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m0s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m10s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m20s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m30s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m40s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [7m50s elapsed]
module.runtime_context.btp_subaccount_environment_instance.kyma[0]: Still creating... [8m0s elapsed]
╷
│ Error: API Error Creating Resource Environment Instance (Subaccount)
│
│   with module.runtime_context.btp_subaccount_environment_instance.kyma[0],
│   on .terraform/modules/runtime_context/btp-context/runtime-context/modules/kyma-runtime/bootstrap-kymaruntime.tf line 258, in resource "btp_subaccount_environment_instance" "kyma":
│  258: resource "btp_subaccount_environment_instance" "kyma" {
│
│ received response with unexpected status [Status: 500; Correlation ID:
│ 97b83bdd-2a7c-8398-866b-ed3d857a46e8]
╵
Operation failed: failed running terraform apply (exit 1)
Error: Process completed with exit code 1.
```

All in all, the outcome is that the kyma runtime provisioning with tf btp provider is hugely unreliable.....regardless of the origin of the errors.

The BTP CLI polling API should not be trusted --> thus, a custom logic of retries, timeouts and circuit breakers should be implemented within the btp provider to address it.

In the meantime, all I can do is to revert to BTP CLI shell scripting and implement my safety net logic there.
Any other way around it ? Please advise.

**Comment by lechnerc77** [comment]
@ptesny I agree that a 504 should be handled by the provider. However, this is the only 50x that makes sense to retry. 
The Terraform provider cannot "heal" a shaky foundation. So we will also discuss with the BTP CLI team how to improve the overall situation and as far as I know there a plans concerning this overall topic.
In addition a creation tije for a K8s runtime that takes more than 20 minutes is also something that needs to be discussed internally.
Until the provider is enhanced to handle 504 gracefully, there is nothing you can do from a consumer side to improve the situation.
